ext/storage_sources/gcp_store/gcp_storage_source.cpp

```diff
@@ -48,42 +57,162 @@ static int gcp_remove(WT_FILE_SYSTEM *, WT_SESSION *, const char *, uint32_t)
+        return EINVAL;
+    }
+
+    gcp_store *gcp = reinterpret_cast<gcp_store *>(storage_source);
+    int ret;
+
+    // Get any prefix to be used for the object keys.
+    WT_CONFIG_ITEM obj_prefix_conf;
+    std::string obj_prefix;
+    if ((ret = gcp->wt_api->config_get_string(
+           gcp->wt_api, session, config, "prefix", &obj_prefix_conf)) == 0)
+        obj_prefix = std::string(obj_prefix_conf.str, obj_prefix_conf.len);
+    else if (ret != WT_NOTFOUND) {
+        std::cerr << "gcp_customize_fs: error parsing config for object prefix." << std::endl;
+        return ret;
+    }
+
+    // Fetch the native WT file system.
+    WT_FILE_SYSTEM *wt_file_system;
+    if ((ret = gcp->wt_api->file_system_get(gcp->wt_api, session, &wt_file_system)) != 0) {
+        std::cerr << "gcp_customize_fs: failed to fetch the native WT file system" << std::endl;
+        return ret;
+    }
+
+    // Get a copy of the home directory.
+    const std::string home_dir = session->connection->get_home(session->connection);
+
+    // Create the file system.
+    // Allocate memory for file system.
+    gcp_file_system *fs;
+    if ((fs = (gcp_file_system *)calloc(1, sizeof(gcp_file_system))) == nullptr) {
+        std::cerr << "gcp_customize_fs: unable to allocate memory for file system." << std::endl;
+        return ENOMEM;
+    }
+
+    // Set variables specific to GCP.
+    fs->storage_source = gcp;
+    fs->wt_file_system = wt_file_system;
+    fs->home_dir = home_dir;
+
+    // Create connection to google cloud.
+    try {
+        fs->gcp_conn = std::make_unique<gcp_connection>(bucket, obj_prefix);
+    } catch (std::invalid_argument &e) {
+        std::cerr << "gcp_customize_file_system: " << e.what() << std::endl;
+        return EINVAL;
+    }
+
+    // Map google cloud functions to the file system.
+    fs->file_system.fs_directory_list = gcp_object_list;
+    fs->file_system.fs_directory_list_single = gcp_object_list_single;
+    fs->file_system.fs_directory_list_free = gcp_object_list_free;
+    fs->file_system.terminate = gcp_file_system_terminate;
+    fs->file_system.fs_exist = gcp_file_exists;
+    fs->file_system.fs_open_file = gcp_file_open;
+    fs->file_system.fs_remove = gcp_remove;
+    fs->file_system.fs_rename = gcp_rename;
+    fs->file_system.fs_size = gcp_file_size;
+
+    // Add to the list of the active file systems. Lock will be freed when the scope is exited.
+    {
+        std::lock_guard<std::mutex> lock_guard(gcp->fs_list_mutex);
+        gcp->fs_list.push_back(fs);
+    }
+
+    *file_system = &fs->file_system;
 
     return 0;
 }
 
 static int
-gcp_add_reference(WT_STORAGE_SOURCE *store)
+gcp_add_reference(WT_STORAGE_SOURCE *storage_source)
 {
-    WT_UNUSED(store);
+    gcp_store *gcp = reinterpret_cast<gcp_store *>(storage_source);
+
+    if (gcp->reference_count == 0 || gcp->reference_count + 1 == 0) {
+        std::cerr << "gcp_add_reference: missing reference or overflow." << std::endl;
```

Comment:
Would it be useful to split this into two cases? A missing reference is a lot different than overflowing the reference count!